### ISSUE TITLE
[MJSONWP] Encountered internal error running command: Error: Command 'java -version' exited with code 20

### ISSUE THREAD

**Issue by vaznoe** [appium/appium#9591]
## The problem

I got this error: [MJSONWP] Encountered internal error running command: Error: Command 'java -version' exited with code 20

## Environment

* Appium version (or git revision) that exhibits the issue: 1.7.1
* Last Appium version that did not exhibit the issue (if applicable):
* Desktop OS/version used to run Appium: macOS Sierra 10.12.6
* Node.js version (unless using Appium.app|exe): v8.9.0
* Mobile platform/version under test: Nexus 5
* Real device or emulator/simulator: emulator
* Appium CLI or Appium.app|exe: 

## Details

If necessary, describe the problem you have been experiencing in more detail.

## Link to Appium logs
```
name-MacBook-Pro:~ name$ appium &
[1] 12589
name-MacBook-Pro:~ name$ [Appium] Welcome to Appium v1.7.1
[Appium] Appium REST http interface listener started on 0.0.0.0:4723
[HTTP] --> POST /wd/hub/session {"desiredCapabilities":{"browserName":"chrome","deviceName":"emulator-5554","platformName":"Android","platformVersion":"6.0"},"capabilities":{"desiredCapabilities":{"browserName":"chrome","deviceName":"emulator-5554","platformName":"Android","platformVersion":"6.0"},"firstMatch":[{"browserName":"chrome","platformName":"android"}]}}
[debug] [MJSONWP] Calling AppiumDriver.createSession() with args: [{"browserName":"chrome","deviceName":"emulator-5554","platformName":"Android","platformVersion":"6.0"},null,{"desiredCapabilities":{"browserName":"chrome","deviceName":"emulator-5554","platformName":"Android","platformVersion":"6.0"},"firstMatch":[{"browserName":"chrome","platformName":"android"}]}]
[debug] [BaseDriver] Event 'newSessionRequested' logged at 1509992833473 (10:27:13 GMT-0800 (PST))
[Appium] Creating new AndroidDriver (v1.28.1) session
[Appium] Capabilities:
[Appium]   browserName: 'chrome'
[Appium]   deviceName: 'emulator-5554'
[Appium]   platformName: 'Android'
[Appium]   platformVersion: '6.0'
[BaseDriver] Session created with session id: 48f1d66a-7487-46e0-ae88-f4a49e4b1117
[debug] [AndroidDriver] Getting Java version
[debug] [AndroidDriver] Shutting down Android driver
[debug] [AndroidDriver] Called deleteSession but bootstrap wasn't active
[MJSONWP] Encountered internal error running command: Error: Command 'java -version' exited with code 20
    at ChildProcess.<anonymous> (../../lib/teen_process.js:92:19)
    at emitTwo (events.js:126:13)
    at ChildProcess.emit (events.js:214:7)
    at maybeClose (internal/child_process.js:925:16)
    at Socket.stream.socket.on (internal/child_process.js:346:11)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at Pipe._handle.close [as _onclose] (net.js:554:12)
[HTTP] <-- POST /wd/hub/session 500 51 ms - 180 
```

## Code To Reproduce Issue [ Good To Have ]
```java
   protected AndroidDriver driver;
    AppiumDriver driver;
    protected String deviceName = "192.168.00.000:5555";
//    protected String deviceName = "emulator-5554";

    public void setUp() {
        DesiredCapabilities capabilities = new DesiredCapabilities();
        capabilities.setCapability("browserName", "chrome");
        capabilities.setCapability("deviceName", deviceName);
        capabilities.setCapability("platformName", "Android");
        capabilities.setCapability("platformVersion","6.0" );
        try {
            driver = new AndroidDriver(new URL("http://127.0.0.1:4723/wd/hub"), capabilities);
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }
    }

    @Test
    public void openMainPage() {
        driver.get("https://......");
    }
```

**Comment by imurchie** [comment]
What is the output of `java -version` when run manually on your system?

**Comment by vaznoe** [comment]
1.8

**Comment by imurchie** [comment]
And what is the return code from that operation?

**Comment by vaznoe** [comment]
objc[5356]: Class JavaLaunchHelper is implemented in both /Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/bin/java (0x107cd64c0) and /Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/lib/libinstrument.dylib (0x107d9e4e0). One of the two will be used. Which one is undefined.

org.openqa.selenium.WebDriverException: An unknown server-side error occurred while processing the command. Original error: Command 'java -version' exited with code 20 (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 214 milliseconds
Build info: version: '3.7.0', revision: '2321c73', time: '2017-11-02T22:22:35.584Z'
System info: host: 'Tatianas-MacBook-Pro.local', ip: 'fe80:0:0:0:8e8:664a:b06f:7c2%en0', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.12.6', java.version: '1.8.0_151'
Driver info: driver.version: AndroidDriver

	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:214)
	at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:166)
	at org.openqa.selenium.remote.JsonWireProtocolResponse.lambda$new$0(JsonWireProtocolResponse.java:53)
	at org.openqa.selenium.remote.JsonWireProtocolResponse.lambda$getResponseFunction$2(JsonWireProtocolResponse.java:91)
	at org.openqa.selenium.remote.ProtocolHandshake.lambda$createSession$0(ProtocolHandshake.java:123)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:958)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:464)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:126)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:73)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:142)
	at io.appium.java_client.remote.AppiumCommandExecutor.execute(AppiumCommandExecutor.java:89)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:600)
	at io.appium.java_client.DefaultGenericMobileDriver.execute(DefaultGenericMobileDriver.java:42)
	at io.appium.java_client.AppiumDriver.execute(AppiumDriver.java:1)
	at io.appium.java_client.android.AndroidDriver.execute(AndroidDriver.java:1)
	at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:219)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:142)
	at io.appium.java_client.DefaultGenericMobileDriver.<init>(DefaultGenericMobileDriver.java:38)
	at io.appium.java_client.AppiumDriver.<init>(AppiumDriver.java:83)
	at io.appium.java_client.AppiumDriver.<init>(AppiumDriver.java:93)
	at io.appium.java_client.android.AndroidDriver.<init>(AndroidDriver.java:72)
	at Tests.setUp(Tests.java:34)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:108)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:523)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:224)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:146)
	at org.testng.internal.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:166)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:105)
	at org.testng.TestRunner.privateRun(TestRunner.java:744)
	at org.testng.TestRunner.run(TestRunner.java:602)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:380)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:375)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1301)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1226)
	at org.testng.TestNG.runSuites(TestNG.java:1144)
	at org.testng.TestNG.run(TestNG.java:1115)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:123)


Test ignored.

===============================================
Default Suite
Total tests run: 1, Failures: 0, Skips: 1
Configuration Failures: 1, Skips: 0
===============================================


Process finished with exit code 0

**Comment by vaznoe** [comment]
resolved. thank u

**Comment by victor61zhong** [comment]
@vaznoe  may I ask how you resolved this issue? I meet the same problem. 
Before this, I did not set JAVA_HOME in my bash_profile. But my script can be run at that moment.
Once I set JAVA_HOME in my bash_profile, my script failed. "selenium.common.exceptions.WebDriverException: Message: An unknown server-side error occurred while processing the command. Original error: Command 'java -version' exited with code 20"

I am using python to write appium script, thx.
